client-v2/src/util/storeMiddleware.js

```diff
@@ -27,3 +32,109 @@ export const updateStateFromUrlChange: Middleware<State, Action> = ({
 
   return result;
 };
+
+type PersistCollectionMeta = {|
+  meta: {
+    // The resource to persist the data to
+    persistTo: 'collection',
+    // The key to take from the action payload. Defaults to 'id'.
+    key?: string,
+    // Should we find collection parents before or after the reducer is called?
+    // This is important when the relevant collection is affected by when the operation
+    // occurs - finding the parent collection before a remove operation, for example,
+    // or after an add operation.
+    applyBeforeReducer?: boolean
+  }
+|};
+
+/**
+ * Return an array of actions - either a single action,
+ * or multiple actions if the action contains batched actions.
+ */
+const unwrapBatchedActions = (action: ActionWithBatchedActions): Action[] =>
+  action.type === 'BATCHING_REDUCER.BATCH' ? action.payload : [action];
+
+/**
+ * Watches for actions that require a collection update, finds the relevant
+ * collection, and dispatches the appropriate action.
+ *
+ * There's quite a lot of code here that's specific to redux-batched-actions,
+ * which would be nice to remove in favour of a more declarative library if
+ * possible.
+ */
+const persistCollectionOnEdit: (
+  (collection: Collection) => Action | ThunkAction
+) => Middleware<Store, Action> = (
+  updateCollectionAction: (
+    collection: Collection
+  ) => Action | ThunkAction = updateCollection
+) => store => {
+  /**
+   * Get the relevant collection ids for the given actions.
+   * @todo At the moment this just cares about updates to article fragments,
+   *   but it should also listen for edits to collections.
+   */
+  const getCollectionIdsForActions = (actions: Action[]) => {
+    const articleFragmentIds: string[] = uniq(
+      actions.map(
+        // A sneaky 'any' here, as it's difficult to handle dynamic key
+        // values with static action types.
+        (act: any) =>
+          act.payload && act.meta.persistTo && act.meta.key
```

Comment:
How come we're checking for `persistTo` here - should it not always be set?

Reply:
redux-batched-actions provides a meta key too, so we must still check for the property.